src/rt/sections_elf_shared.d

```diff
@@ -711,12 +711,12 @@ void scanSegments(in ref dl_phdr_info info, DSO* pdso)
         case PT_LOAD:
             if (phdr.p_flags & PF_W) // writeable data segment
             {
-                auto beg = cast(void*)(info.dlpi_addr + phdr.p_vaddr);
+                auto beg = cast(void*)(info.dlpi_addr + (phdr.p_vaddr & ~(size_t.sizeof - 1)));
                 pdso._gcRanges.insertBack(beg[0 .. phdr.p_memsz]);
             }
             version (Shared) if (phdr.p_flags & PF_X) // code segment
             {
-                auto beg = cast(void*)(info.dlpi_addr + phdr.p_vaddr);
+                auto beg = cast(void*)(info.dlpi_addr + (phdr.p_vaddr & ~(size_t.sizeof - 1)));
                 pdso._codeSegments.insertBack(beg[0 .. phdr.p_memsz]);
```

Comment:
I don't think that's quite correct.
- Only the address space between `[dlpi_addr .. p_vaddr+p_memsz]` is accessible.  Attempting to read outside this results in a segfault (at least, in our hardened CI environment)
- The end address should be on a pointer aligned boundary, and if it isn't, we don't want to read it - else we'll end up segfaulting in the GC.

In your example, your "correct" case would do the following:
1. Set `codeSegments` length to 6
2. GC will scan the first pointer `[0 .. 4]`, move to the next pointer position.
3. GC will scan the second pointer `[4 .. 8]` - oops, out of bounds.